Error while checking build/O2/fullCI for 97c7bab36985d52afee631ad49aab4bcf004af11 at 2023-02-21 06:43:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/sm_20_atomic_functions.h:103:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10792-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/sm_20_atomic_functions.h:103:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: expected expression [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: type name does not allow function specifier to be specified [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: use of undeclared identifier 'noinline'; did you mean 'inline'? [clang-diagnostic-error]
/O2/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10792/97c7bab36985d52afee631ad49aab4bcf004af11/build_O2_fullCI/pretty.html).

--- a/Detectors/TRD/qc/include/TRDQC/Tracking.h
+++ b/Detectors/TRD/qc/include/TRDQC/Tracking.h
@@ -18,6 +18,7 @@
 
 #include "DataFormatsTRD/TrackTRD.h"
 #include "DataFormatsTRD/Tracklet64.h"
+#include "TRDBase/PadCalibrationsAliases.h"
 #include "DataFormatsTRD/CalibratedTracklet.h"
 #include "DataFormatsTRD/Constants.h"
 #include "ReconstructionDataFormats/TrackTPCITS.h"
@@ -25,9 +26,9 @@
 #include "DataFormatsTPC/TrackTPC.h"
 #include "DetectorsBase/Propagator.h"
 #include "TRDBase/RecoParam.h"
-#include "TH1.h"
 
 #include "Rtypes.h"
+#include "TH1.h"
 
 #include <gsl/span>
 #include <bitset>
@@ -83,12 +84,14 @@ struct TrackQC {
   std::array<int, constants::NLAYER> trackletPositionSigned{}; ///< the raw position from Tracklet64 (signed integer)
   std::array<int, constants::NLAYER> trackletDet{};            ///< the chamber of the tracklet
   // some tracklet details to identify its global MCM number to check if it is from noisy MCM
-  std::array<int, constants::NLAYER> trackletHCId{};                                     ///< the half-chamber ID of the tracklet
-  std::array<int, constants::NLAYER> trackletRob{};                                      ///< the ROB number of the tracklet
-  std::array<int, constants::NLAYER> trackletMcm{};                                      ///< the MCM number of the tracklet
-  std::array<float, constants::NLAYER> trackletChi2{};                                   ///< estimated chi2 for the update of the track with the given tracklet
-  std::array<std::array<int, constants::NCHARGES>, constants::NLAYER> trackletCharges{}; ///< charges of tracklets
-  ClassDefNV(TrackQC, 4);
+  std::array<int, constants::NLAYER> trackletHCId{};                                          ///< the half-chamber ID of the tracklet
+  std::array<int, constants::NLAYER> trackletRob{};                                           ///< the ROB number of the tracklet
+  std::array<int, constants::NLAYER> trackletMcm{};                                           ///< the MCM number of the tracklet
+  std::array<float, constants::NLAYER> trackletChi2{};                                        ///< estimated chi2 for the update of the track with the given tracklet
+  std::array<std::array<int, constants::NCHARGES>, constants::NLAYER> trackletCharges{};      ///< charges of tracklets
+  std::array<std::array<float, constants::NCHARGES>, constants::NLAYER> trackletCorCharges{}; ///< corrected charges of tracklets
+
+  ClassDefNV(TrackQC, 5);
 };
 
 class Tracking
@@ -118,13 +121,21 @@ class Tracking
   // Make output accessible to DPL processor
   std::vector<TrackQC>& getTrackQC() { return mTrackQC; }
 
+  // Set the local gain factors with values from the ccdb
+  void setLocalGainFactors(const o2::trd::LocalGainFactor& localGain)
+  {
+    mLocalGain = localGain;
+  }
+
  private:
   float mMaxSnp{o2::base::Propagator::MAX_SIN_PHI};  ///< max snp when propagating tracks
   float mMaxStep{o2::base::Propagator::MAX_STEP};    ///< maximum step for propagation
   MatCorrType mMatCorr{MatCorrType::USEMatCorrNONE}; ///< if material correction should be done
   RecoParam mRecoParam;                              ///< parameters required for TRD reconstruction
+
   // QA results
   std::vector<TrackQC> mTrackQC;
+
   // input from DPL
   gsl::span<const o2::dataformats::TrackTPCITS> mTracksITSTPC; ///< ITS-TPC seeding tracks
   gsl::span<const o2::tpc::TrackTPC> mTracksTPC;               ///< TPC seeding tracks
@@ -133,7 +144,10 @@ class Tracking
   gsl::span<const Tracklet64> mTrackletsRaw;                   ///< array of raw tracklets needed for TRD refit
   gsl::span<const CalibratedTracklet> mTrackletsCalib;         ///< array of calibrated tracklets needed for TRD refit
 
-  ClassDefNV(Tracking, 1);
+  // corrections from ccdb, some need to be loaded only once hence an init flag
+  o2::trd::LocalGainFactor mLocalGain; ///< local gain factors from krypton calibration
+
+  ClassDefNV(Tracking, 2);
 };
 
 } // namespace trd

--- a/Detectors/TRD/qc/src/Tracking.cxx
+++ b/Detectors/TRD/qc/src/Tracking.cxx
@@ -77,7 +77,8 @@ void Tracking::checkTrack(const TrackTRD& trkTrd, bool isTPCTRD)
     if (trkltId < 0) {
       continue;
     }
-    int trkltDet = mTrackletsRaw[trkltId].getDetector();
+    const auto& tracklet = mTrackletsRaw[trkltId];
+    int trkltDet = tracklet.getDetector();
     int trkltSec = trkltDet / (NLAYER * NSTACK);
     if (trkltSec != o2::math_utils::angle2Sector(trk.getAlpha())) {
       if (!trk.rotate(o2::math_utils::sector2Angle(trkltSec))) {
@@ -94,14 +95,14 @@ void Tracking::checkTrack(const TrackTRD& trkTrd, bool isTPCTRD)
     float tilt = tan(TMath::DegToRad() * pad->getTiltingAngle()); // tilt is signed! and returned in degrees
     float tiltCorrUp = tilt * (mTrackletsCalib[trkltId].getZ() - trk.getZ());
     float zPosCorrUp = mTrackletsCalib[trkltId].getZ() + mRecoParam.getZCorrCoeffNRC() * trk.getTgl();
-    float padLength = pad->getRowSize(mTrackletsRaw[trkltId].getPadRow());
+    float padLength = pad->getRowSize(tracklet.getPadRow());
     if (!((trk.getSigmaZ2() < (padLength * padLength / 12.f)) && (std::fabs(mTrackletsCalib[trkltId].getZ() - trk.getZ()) < padLength))) {
       tiltCorrUp = 0.f;
     }
 
     std::array<float, 2> trkltPosUp{mTrackletsCalib[trkltId].getY() - tiltCorrUp, zPosCorrUp};
     std::array<float, 3> trkltCovUp;
-    mRecoParam.recalcTrkltCov(tilt, trk.getSnp(), pad->getRowSize(mTrackletsRaw[trkltId].getPadRow()), trkltCovUp);
+    mRecoParam.recalcTrkltCov(tilt, trk.getSnp(), pad->getRowSize(tracklet.getPadRow()), trkltCovUp);
     auto chi2trklt = trk.getPredictedChi2(trkltPosUp, trkltCovUp);
 
     qcStruct.trackX[iLayer] = trk.getX();
@@ -117,20 +118,43 @@ void Tracking::checkTrack(const TrackTRD& trkTrd, bool isTPCTRD)
     qcStruct.trackletY[iLayer] = trkltPosUp[0];
     qcStruct.trackletZ[iLayer] = trkltPosUp[1];
     qcStruct.trackletDy[iLayer] = mTrackletsCalib[trkltId].getDy();
-    qcStruct.trackletSlope[iLayer] = mTrackletsRaw[trkltId].getSlope();
-    qcStruct.trackletSlopeSigned[iLayer] = mTrackletsRaw[trkltId].getSlopeBinSigned();
-    qcStruct.trackletPosition[iLayer] = mTrackletsRaw[trkltId].getPosition();
-    qcStruct.trackletPositionSigned[iLayer] = mTrackletsRaw[trkltId].getPositionBinSigned();
+    qcStruct.trackletSlope[iLayer] = tracklet.getSlope();
+    qcStruct.trackletSlopeSigned[iLayer] = tracklet.getSlopeBinSigned();
+    qcStruct.trackletPosition[iLayer] = tracklet.getPosition();
+    qcStruct.trackletPositionSigned[iLayer] = tracklet.getPositionBinSigned();
     qcStruct.trackletDet[iLayer] = trkltDet;
-    qcStruct.trackletHCId[iLayer] = mTrackletsRaw[trkltId].getHCID();
-    qcStruct.trackletRob[iLayer] = mTrackletsRaw[trkltId].getROB();
-    qcStruct.trackletMcm[iLayer] = mTrackletsRaw[trkltId].getMCM();
+    qcStruct.trackletHCId[iLayer] = tracklet.getHCID();
+    qcStruct.trackletRob[iLayer] = tracklet.getROB();
+    qcStruct.trackletMcm[iLayer] = tracklet.getMCM();
     qcStruct.trackletChi2[iLayer] = chi2trklt;
     qcStruct.trackletCharges[iLayer] = {
-      mTrackletsRaw[trkltId].getQ0(),
-      mTrackletsRaw[trkltId].getQ1(),
-      mTrackletsRaw[trkltId].getQ2(),
+      tracklet.getQ0(),
+      tracklet.getQ1(),
+      tracklet.getQ2(),
     };
+
+    // Corrected Tracklets
+    auto cor = mLocalGain.getValue(tracklet.getHCID() / 2, tracklet.getPadCol(), tracklet.getPadRow()) * (std::abs(tracklet.getSlopeBinSigned()) + 1.f);
+    float q0{tracklet.getQ0() / cor}, q1{tracklet.getQ1() / cor}, q2{tracklet.getQ2() / cor};
+
+    // z-row merging
+    if (trkTrd.getIsCrossingNeighbor(iLayer) && trkTrd.getHasNeighbor()) {
+      for (const auto& trklt : mTrackletsRaw) {
+        if (tracklet.getTrackletWord() == trklt.getTrackletWord()) { // skip original tracklet
+          continue;
+        }
+        if (std::abs(tracklet.getPadCol() - trklt.getPadCol()) == 1 && std::abs(tracklet.getPadRow() - trklt.getPadRow()) == 0) {
+          // Add charge information
+          auto cor = mLocalGain.getValue(trklt.getHCID() / 2, trklt.getPadCol(), trklt.getPadRow()) * (std::abs(trklt.getSlopeBinSigned()) + 1.f);
+          q0 += trklt.getQ0() / cor;
+          q1 += trklt.getQ1() / cor;
+          q2 += trklt.getQ2() / cor;
+          break;
+        }
+      }
+    }
+
+    qcStruct.trackletCorCharges[iLayer] = {q0, q1, q2};
   }
   mTrackQC.push_back(qcStruct);
 }

--- a/Detectors/TRD/workflow/include/TRDWorkflow/TRDGlobalTrackingQCSpec.h
+++ b/Detectors/TRD/workflow/include/TRDWorkflow/TRDGlobalTrackingQCSpec.h
@@ -23,6 +23,7 @@
 #include "Framework/DataProcessorSpec.h"
 #include "Framework/Task.h"
 #include "Framework/ConfigParamRegistry.h"
+#include "Framework/CCDBParamSpec.h"
 #include "DPLUtils/MakeRootTreeWriterSpec.h"
 #include "DetectorsBase/GeometryManager.h"
 #include "DetectorsBase/Propagator.h"
@@ -71,12 +72,17 @@ class TRDGlobalTrackingQC : public Task
   {
     o2::base::GRPGeomHelper::instance().checkUpdates(pc);
     static bool initOnceDone = false;
-    if (!initOnceDone) { // this params need to be queried only once
+    if (!initOnceDone) { // these params need to be queried only once
       initOnceDone = true;
 
       mQC.init();
+
+      // Local pad gain calibration from krypton run
+      auto localGain = *(pc.inputs().get<o2::trd::LocalGainFactor*>("localgainfactors"));
+      mQC.setLocalGainFactors(localGain);
     }
   }
+
   std::shared_ptr<DataRequest> mDataRequest;
   std::shared_ptr<o2::base::GRPGeomRequest> mGGCCDBRequest;
   Tracking mQC;
@@ -100,6 +106,7 @@ DataProcessorSpec getTRDGlobalTrackingQCSpec(o2::dataformats::GlobalTrackID::mas
 
   dataRequest->requestTracks(src, false);
   dataRequest->requestClusters(srcClu, false);
+  dataRequest->inputs.emplace_back("localgainfactors", "TRD", "LOCALGAINFACTORS", 0, Lifetime::Condition, ccdbParamSpec("TRD/Calib/LocalGainFactor"));
   auto ggRequest = std::make_shared<o2::base::GRPGeomRequest>(false,                             // orbitResetTime
                                                               false,                             // GRPECS=true
                                                               false,                             // GRPLHCIF

